### ISSUE TITLE
[consistent-return] Reports error with exhaustive switches

### ISSUE THREAD

**Issue by renchap** [typescript-eslint/typescript-eslint#4510]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "consistent-return": "error",
    "@typescript-eslint/switch-exhaustiveness-check": "error",
  }
}
```

```TS
type Param = "a" | "b"

function exhaustiveSwitch(p: Param): number {
  switch (p) {
    case "a":
      return 1
    case "b":
      return 2
  }
}
```

**Expected Result**

There are no errors

**Actual Result**

I am getting this error:

```
function exhaustiveSwitch(p: Param): number
               ^^ Expected to return a value at the end of function 'exhaustiveSwitch'  eslint[consistent-return]
```

**Additional Info**

This is because the base `consistent-return` eslint check does not take types into account. It could be augmented by `typescript-eslint` so it knows that all paths are returning something (as with the `switch-exhaustiveness-check` rule).

This might be related to #1277 and #1987

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.2` |
| `@typescript-eslint/parser`        | `5.10.2` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.8.0` |
| `node`                             | `16.13.1` |

**Comment by bradzacher** [comment]
TS should do all the same checks that the rule does, except better.

For example - [here are all of the invalid test cases that ESLint has for the rule](https://www.typescriptlang.org/play?ssl=2&ssc=36&pln=3&pc=1#code/CYUwxgNghgTiAEYD2A7AzgF3lAXPARkkhCFCgNwBQoksCAZgK4pgYCWq89AFLAOZ4yATwCUeAG5I2wKpVpo08AIIBGeAG94AYSgAHNhigQ2ALxDAAYs1YcU3ERvht68Xg7gZGMFPAwxGIOTwAL4hctAKygBMGvRE9o7OrlDuIJ7evv6BIcFU4rDwSPgAVvAAvLHxDppJbvAeXj5+AUHBuZQ8CWUAfIkudQ0ZzdnBIlQ8TCzsqAk1-Sn1aY2ZLSFjHdyTNpxxSCqzTvOp6U1Zretb0z4A+rtRB7ULg6erwR3WV1xEAMwPR4snFYjd5TWxfJAAFj+rmGx2Wkmk8AADEFnkDWiDtj5dgBWaHcWEA5bMUD0NgocyopZDM5hS5g3YANnxhLRwypgIRwGRGPpOyIAHYWVk4TTVmiSSAyRSZHTeOVenMYSKid4giAIGgEGj6EYtecNny7NVDsqAqKKPANVrVdi9SMRIaPrZhebbeirZrtdTLaNMZ9dgAOV0gC3qr3u3VejH5GDggCc5VcDh6fTNofd7M9NuevOd-KQIYtHut3pOrSAA). You can see that TS errors on *all* of them.

For that reason - I wouldn't bother with `consistent-return` in TS code.
I don't think it's worth the effort to create an extension rule for this, given that it doesn't provide any additional checks on top of TS. I'd even go as far as saying that we should turn this off in our recommended config.

**Comment by renchap** [comment]
This makes sense!

I will disable this rule in my config and I think it can safely be done in the recommended config :)

I let you choose to close the issue, or keep it open if you want to track the recommended config change.
